src/vendor/jx/cubes.js

```diff
@@ -392,4 +394,35 @@ function window(cubes, { value, edges: edgeNames, along }) {
   );
 }
 
+/*
+Groupby, but with all combinations of all columns grouped.
+The result is s a cube of lists, where the lists are elements from `self`
+edges.value is used to determine what part of each edge a record belowngs
+For 2 dimensions this is a pivot table, for more dimensions it is a "cube".
+Google "sql group by cube" for more information
+ */
+ArrayWrapper.edges = (self, edges, zero = array) => {
+  const normalizedEdges = edges.map(Edge.newInstance);
+  const dims = normalizedEdges.map(e => e.domain.partitions.length);
+  const matrix = new Matrix({ dims, zero });
+
+  self.forEach(row => {
+    const coord = normalizedEdges.map(e => e.domain.valueToIndex(e.value(row)));
+
+    zip(dims, normalizedEdges).forEach(([d, e], i) => {
+      if (e.domain.type === 'value' && d < e.domain.partitions.length) {
+        // last element of value domain is NULL, ensure it is still last
+        matrix.insertPart(i, d - 1);
+        dims[i] = d + 1;
+      }
+    });
+
+    matrix.add(coord, row);
+  });
+
+  normalizedEdges.forEach(e => e.domain.lock());
+
+  return new Cube(normalizedEdges, matrix);
+};
+
```

Comment:
Moved from elsewhere: Tests did not pass because of yet-another-import-loop.  We break the loop by adding static class methods to already-imported classes.